common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolver.java

```diff
@@ -179,6 +179,18 @@ public ManagedBuffer getBlockData(
     return getSortBasedShuffleBlockData(executor, shuffleId, mapId, reduceId);
   }
 
+  public ManagedBuffer getBlockData(
```

Comment:
There's code elsewhere in this class (`deleteNonShuffleFiles`) that seems to be for deleting non-shuffle files when executors die. I wonder why that is not causing problems for you.

A quick look seems to tell me it's hooked up only in standalone mode, which means two things:
- it should probably be wired up in YARN also
- it should be fixed so that it doesn't cause problems with the code you're adding

Reply:
The call chain lead to the Worker (and I have not found any other place where this is called):
https://github.com/apache/spark/blob/df3a80da4270c7b5eddb83383d8149ed64b25a66/core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala#L753-L755

I can see several possible solutions here:
1) mixing "spark.shuffle.service.enabled" into this condition so basically switching off the cleanup when external shuffle service is enabled and producing just a warning about this decision and the reason behind.
2) add a `require` (so throw an `IllegalArgumentException`) when both "spark.storage.cleanupFilesAfterExecutorExit" and external shuffle service is enabled.

As I see standalone mode already has some weaknesses in this area:
https://github.com/apache/spark/blob/df3a80da4270c7b5eddb83383d8149ed64b25a66/core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala#L779-L791

I am open for more possibilities. Do you agree the first one with the warning is good solution here?

Reply:
I have analysed this problem a little more: external shuffle service might do the cleanup of executors' local directories when the application is removed. It is done by calling org.apache.spark.network.shuffle.ExternalShuffleBlockResolver#applicationRemoved:

https://github.com/apache/spark/blob/df3a80da4270c7b5eddb83383d8149ed64b25a66/common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolver.java#L221-L226

My uncertainty comes from this condition which is on the call chain:

https://github.com/apache/spark/blob/df3a80da4270c7b5eddb83383d8149ed64b25a66/core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala#L637-L639 

In case of `ApplicationFinished` the first part is definitely true but I have to check the second one a bit more and even do a test. So I will analyse this further and I will come back to you with my findings.

Reply:
I have tried to reproduce the problem and meanwhile I think you are absolutely right and as non-shuffle files are delete recompute must be triggered. But in my test somehow the blocks are still found.

Log parts (with 7 partitions as I have 8 cores in this machine). Here are the cleanups in the worker log:

```
19/05/01 15:06:11 INFO ExternalShuffleBlockResolver: Cleaning up non-shuffle files in executor AppExecId{appId=app-20190501150515-0001, execId=5}'s 1 local dirs
19/05/01 15:06:11 INFO Worker: Executor app-20190501150515-0001/4 finished with state KILLED exitStatus 143
19/05/01 15:06:11 INFO ExternalShuffleBlockResolver: Clean up non-shuffle files associated with the finished executor 4
19/05/01 15:06:11 INFO ExternalShuffleBlockResolver: Cleaning up non-shuffle files in executor AppExecId{appId=app-20190501150515-0001, execId=4}'s 1 local dirs
19/05/01 15:06:11 INFO Worker: Executor app-20190501150515-0001/6 finished with state KILLED exitStatus 143
19/05/01 15:06:11 INFO ExternalShuffleBlockResolver: Clean up non-shuffle files associated with the finished executor 6
19/05/01 15:06:11 INFO ExternalShuffleBlockResolver: Cleaning up non-shuffle files in executor AppExecId{appId=app-20190501150515-0001, execId=6}'s 1 local dirs
19/05/01 15:06:11 INFO Worker: Executor app-20190501150515-0001/3 finished with state KILLED exitStatus 143
19/05/01 15:06:11 INFO ExternalShuffleBlockResolver: Clean up non-shuffle files associated with the finished executor 3
19/05/01 15:06:11 INFO ExternalShuffleBlockResolver: Cleaning up non-shuffle files in executor AppExecId{appId=app-20190501150515-0001, execId=3}'s 1 local dirs
19/05/01 15:06:11 INFO Worker: Executor app-20190501150515-0001/7 finished with state KILLED exitStatus 143
19/05/01 15:06:11 INFO ExternalShuffleBlockResolver: Clean up non-shuffle files associated with the finished executor 7
19/05/01 15:06:11 INFO ExternalShuffleBlockResolver: Cleaning up non-shuffle files in executor AppExecId{appId=app-20190501150515-0001, execId=7}'s 1 local dirs
19/05/01 15:06:14 INFO Worker: Asked to kill executor app-20190501150515-0001/0
19/05/01 15:06:14 INFO ExecutorRunner: Runner thread for executor app-20190501150515-0001/0 interrupted
19/05/01 15:06:14 INFO ExecutorRunner: Killing process!
19/05/01 15:06:14 INFO Worker: Executor app-20190501150515-0001/0 finished with state KILLED exitStatus 143
19/05/01 15:06:14 INFO ExternalShuffleBlockResolver: Clean up non-shuffle files associated with the finished executor 0
19/05/01 15:06:14 INFO ExternalShuffleBlockResolver: Cleaning up non-shuffle files in executor AppExecId{appId=app-20190501150515-0001, execId=0}'s 1 local dirs
19/05/01 15:06:18 INFO Worker: Asked to kill executor app-20190501150515-0001/2
19/05/01 15:06:18 INFO ExecutorRunner: Runner thread for executor app-20190501150515-0001/2 interrupted
19/05/01 15:06:18 INFO ExecutorRunner: Killing process!
19/05/01 15:06:18 INFO Worker: Executor app-20190501150515-0001/2 finished with state KILLED exitStatus 143
19/05/01 15:06:18 INFO ExternalShuffleBlockResolver: Clean up non-shuffle files associated with the finished executor 2
19/05/01 15:06:18 INFO ExternalShuffleBlockResolver: Cleaning up non-shuffle files in executor AppExecId{appId=app-20190501150515-0001, execId=2}'s 1 local dirs
```

And this part is from the end of my test where persisted RDD referenced again. To execute these tasks a new executor is started. This is from the stderr of that executor:

```
9/05/01 15:06:52 INFO BlockManager: Found block rdd_1_0 remotely
19/05/01 15:06:52 INFO Executor: Finished task 0.0 in stage 2.0 (TID 11). 697 bytes result sent to driver
19/05/01 15:06:52 INFO CoarseGrainedExecutorBackend: Got assigned task 12
19/05/01 15:06:52 INFO Executor: Running task 1.0 in stage 2.0 (TID 12)
19/05/01 15:06:52 INFO BlockManager: Found block rdd_1_1 remotely
19/05/01 15:06:52 INFO Executor: Finished task 1.0 in stage 2.0 (TID 12). 697 bytes result sent to driver
19/05/01 15:06:52 INFO CoarseGrainedExecutorBackend: Got assigned task 13
19/05/01 15:06:52 INFO Executor: Running task 2.0 in stage 2.0 (TID 13)
19/05/01 15:06:52 INFO BlockManager: Found block rdd_1_2 remotely
19/05/01 15:06:52 INFO Executor: Finished task 2.0 in stage 2.0 (TID 13). 697 bytes result sent to driver
19/05/01 15:06:52 INFO CoarseGrainedExecutorBackend: Got assigned task 14
19/05/01 15:06:52 INFO Executor: Running task 3.0 in stage 2.0 (TID 14)
19/05/01 15:06:52 INFO BlockManager: Found block rdd_1_3 remotely
19/05/01 15:06:52 INFO Executor: Finished task 3.0 in stage 2.0 (TID 14). 697 bytes result sent to driver
19/05/01 15:06:52 INFO CoarseGrainedExecutorBackend: Got assigned task 15
19/05/01 15:06:52 INFO Executor: Running task 4.0 in stage 2.0 (TID 15)
19/05/01 15:06:52 INFO BlockManager: Found block rdd_1_4 remotely
19/05/01 15:06:52 INFO Executor: Finished task 4.0 in stage 2.0 (TID 15). 697 bytes result sent to driver
19/05/01 15:06:52 INFO CoarseGrainedExecutorBackend: Got assigned task 16
19/05/01 15:06:52 INFO Executor: Running task 5.0 in stage 2.0 (TID 16)
19/05/01 15:06:52 INFO BlockManager: Found block rdd_1_5 remotely
19/05/01 15:06:52 INFO Executor: Finished task 5.0 in stage 2.0 (TID 16). 697 bytes result sent to driver
19/05/01 15:06:52 INFO CoarseGrainedExecutorBackend: Got assigned task 17
19/05/01 15:06:52 INFO Executor: Running task 6.0 in stage 2.0 (TID 17)
19/05/01 15:06:52 INFO BlockManager: Found block rdd_1_6 remotely
19/05/01 15:06:52 INFO Executor: Finished task 6.0 in stage 2.0 (TID 17). 697 bytes result sent to driver
19/05/01 15:06:52 INFO CoarseGrainedExecutorBackend: Got assigned task 18
19/05/01 15:06:52 INFO Executor: Running task 7.0 in stage 2.0 (TID 18)
19/05/01 15:06:52 INFO BlockManager: Found block rdd_1_7 remotely
```
 
I will come back to this tomorrow. I would like to understand how this works exactly. Otherwise a possible correction (switching off the cleanup) would duplicate these blocks.

Reply:
I have found the problem: `file.length()` is 0 not only for empty files but for not existing ones too. I have a solution but I would like to cover this case with a unit test too.

Reply:
In my next commit three changes regarding the handling of deleted files:

### Short overview of the changes

1) In `ExternalShuffleBlockResolver#getDiskPersistedRddBlockData` file existence is explicitly checked before creating the `FileSegmentManagedBuffer` and if the file is missing `null` is passed. Then extended `OneForOneStreamManager` and `ChunkFetchRequestHandler` to handle correctly this `null` value (in `ChunkFetchRequestHandler` an `IllegalStateException` with the `Chunk was not found` message is thrown and this appears in the logs). 
2) As the previous point is still not 100% perfect there is a change in `BlockManager`. At block fetching the buffer size is required to be non-zero if the blockSize is non zero (blockSize comes from the block manager master). I am afraid here equality between these two properties cannot be checked as a MEMORY_ONLY block might be fetched into the disk.
3) In the `Worker` I have added the warning about the overriding / switching off the `spark.storage.cleanupFilesAfterExecutorExit` setting. I have checked manually (and also the source code) that the cleanup is really done at the stopping of the application.

### Manual tests

Several manual tests are done.

#### testing point 1)

After the 1) point `ExternalShuffleBlockResolver#getDiskPersistedRddBlockData` and related changes (here the my expectation was to see a re-trigger of the calculations over silently fetched empty blocks):

~~~
$ grep "Attila: calculate"  work/app-20190502115425-0000/*/stdout |wc -l
     200

$ grep "Failed to fetch" -A3  work/app-20190502115425-0000/*/stderr
work/app-20190502115425-0000/10/stderr:19/05/02 11:56:07 ERROR RetryingBlockFetcher: Failed to fetch block rdd_1_2, and will not retry (0 retries)
work/app-20190502115425-0000/10/stderr-org.apache.spark.network.client.ChunkFetchFailureException: Failure while fetching StreamChunkId{streamId=1537044795002, chunkIndex=0}: java.lang.IllegalStateException: Chunk was not found
work/app-20190502115425-0000/10/stderr-	at org.apache.spark.network.server.ChunkFetchRequestHandler.channelRead0(ChunkFetchRequestHandler.java:95)
work/app-20190502115425-0000/10/stderr-	at org.apache.spark.network.server.ChunkFetchRequestHandler.channelRead0(ChunkFetchRequestHandler.java:51)
--
work/app-20190502115425-0000/10/stderr:19/05/02 11:56:07 WARN BlockManager: Failed to fetch block after 1 fetch failures. Most recent failure cause:
work/app-20190502115425-0000/10/stderr-org.apache.spark.SparkException: Exception thrown in awaitResult:
work/app-20190502115425-0000/10/stderr-	at org.apache.spark.util.ThreadUtils$.awaitResult(ThreadUtils.scala:225)
work/app-20190502115425-0000/10/stderr-	at org.apache.spark.network.BlockTransferService.fetchBlockSync(BlockTransferService.scala:121)
~~~ 

#### testing point 2)

The `BlockManager` changes also tested separately from the previous point (this also re-triggers the calculations on its own):

~~~
$ grep "Attila: calculate" work/app-20190502100445-0000/*/stdout  | wc -l
     200

$ grep "Failed to fetch" -A2  work/app-20190502110418-0000/*/stderr
work/app-20190502110418-0000/10/stderr:19/05/02 11:06:04 WARN BlockManager: Failed to fetch block after 1 fetch failures. Most recent failure cause:
work/app-20190502110418-0000/10/stderr-java.lang.IllegalStateException: Empty buffer received for non empty block
work/app-20190502110418-0000/10/stderr-	at org.apache.spark.storage.BlockManager.getRemoteManagedBuffer(BlockManager.scala:904)
--
work/app-20190502110418-0000/11/stderr:19/05/02 11:06:04 WARN BlockManager: Failed to fetch block after 1 fetch failures. Most recent failure cause:
work/app-20190502110418-0000/11/stderr-java.lang.IllegalStateException: Empty buffer received for non empty block
work/app-20190502110418-0000/11/stderr-	at org.apache.spark.storage.BlockManager.getRemoteManagedBuffer(BlockManager.scala:904)
--
work/app-20190502110418-0000/12/stderr:19/05/02 11:06:07 WARN BlockManager: Failed to fetch block after 1 fetch failures. Most recent failure cause:
work/app-20190502110418-0000/12/stderr-java.lang.IllegalStateException: Empty buffer received for non empty block
~~~

#### testing together

Finally I have tested the whole solution:

~~~
$ grep "Attila" work/app-20190502202206-0000/*/stdout | wc -l
     100
~~~~

Along the test app is changed to do a local reduce on the collected RDD:

```scala    
println("Calling collect on the first RDD: " + rdd.collect().reduce(_ + _))
``` 

This can be found in the logs:

```
Attila: sum: 4950
```